components/places/ios/Places/LoginRecord.swift

```diff
@@ -0,0 +1,182 @@
+    /// result in an `LoginsStoreError.InvalidLogin`.
+    public var httpRealm: String? = nil
+
+    /// The submission URL for the form where this login may be entered.
+    ///
+    /// As mentioned above, exactly one of `httpRealm` or `formSubmitURL` is allowed
+    /// to be present, and attempting to insert or update a record to have
+    /// both or neither will result in an `LoginsStoreError.InvalidLogin`.
+    public var formSubmitURL: String? = nil
+
+    /// A lower bound on the number of times this record has been "used".
+    ///
+    /// A use is recorded (and `timeLastUsed` is updated accordingly) in
+    /// the following scenarios:
+    ///
+    /// - Newly inserted records have 1 use.
+    /// - Updating a record locally (that is, updates that occur from a
+    ///   sync do not count here) increments the use count.
+    /// - Calling `touch` on the corresponding id.
+    ///
+    /// This is ignored by `add` and `update`.
+    public var timesUsed: Int = 0
+
+    /// An upper bound on the time of creation in milliseconds from the unix epoch.
+    ///
+    /// This is ignored by `add` and `update`.
+    public var timeCreated: Int64 = 0
+
+    /// A lower bound on the time of last use in milliseconds from the unix epoch.
+    ///
+    /// This is ignored by `add` and `update`.
+    public var timeLastUsed: Int64 = 0
+
+    /// A lower bound on the time of last use in milliseconds from the unix epoch.
+    ///
+    /// This is ignored by `add` and `update`.
+    public var timePasswordChanged: Int64 = 0
+
+    /// HTML field name of the username, if known.
+    public var usernameField: String? = nil
+
+    /// HTML field name of the password, if known.
+    public var passwordField: String? = nil
+
+    open func toJSONDict() -> [String: Any] {
+        var dict: [String: Any] = [
+            "id": self.id,
+            "password": self.password,
+            "hostname": self.hostname,
+
+            "timesUsed": self.timesUsed,
+            "timeCreated": self.timeCreated,
+            "timeLastUsed": self.timeLastUsed,
+            "timePasswordChanged": self.timePasswordChanged,
+        ]
+
+        if let username = self.username {
+            dict["username"] = username
+        }
+
+        if let httpRealm = self.httpRealm {
+            dict["httpRealm"] = httpRealm
+        }
+
+        if let formSubmitURL = self.formSubmitURL {
+            dict["formSubmitURL"] = formSubmitURL
+        }
+
+        if let passwordField = self.passwordField {
+            dict["passwordField"] = passwordField
+        }
+
+        if let usernameField = self.usernameField {
+            dict["usernameField"] = usernameField
+        }
+        return dict
+    }
+
+    open func toJSON() throws -> String {
+        // We need a String to pass back to rust.
+        let data: Data = try JSONSerialization.data(withJSONObject: self.toJSONDict())
+        return String(data: data, encoding: String.Encoding.utf8)!
```

Comment:
Consider avoiding the force-unwrap
`if let result = String(...) { return result } else { throw NSError(domain: "toJSON() String() failure", code: 1, userInfo: nil ) }`